[VectorCombine] Allow shuffling with bitcast for not multiple offset for loadsize

ParkHanbum opened the pull request:
Previously, vectorization for load-insert failed when the Offset was not
a multiple of the Load type size.

This patch allow it in two steps,
1. Vectorize it using a common multiple of Offset and LoadSize.
2. Bitcast to fit

Alive2: https://alive2.llvm.org/ce/z/Kgr9HQ

ParkHanbum committed:
[VectorCombine] Allow shuffling with bitcast for not multiple offset for loadsize

Previously, vectorization for load-insert failed when the Offset was not
a multiple of the Load type size.

This patch allow it in two steps,
1. Vectorize it using a common multiple of Offset and LoadSize.
2. Bitcast to fit

Alive2: https://alive2.llvm.org/ce/z/Kgr9HQ

ParkHanbum committed:
add test-cases

ParkHanbum commented:
@RKSimon could you review this patch please?

RKSimon reviewed (commented):
Does this work with big endian?

ParkHanbum commented:
@RKSimon Probably not. Is it possible to separate logic according to Endian?

RKSimon commented:
Yes - use `DL->isBigEndian()`

ParkHanbum commented:
@RKSimon 
Last time, I asked if the number of IRs increases, can we see if it improves the performance? I was able to see it through the number of Assembly. Thank you.

This time, it's a little difficult, so I'm asking you a question.
I understand that CostModel judge the costs internally, but, 
Is there any way to infer performance improvement through assembly?

https://alive2.llvm.org/ce/z/2gqJjB

ParkHanbum committed:
BigEndian check update

ParkHanbum committed:
use std::gcd instead self code

ParkHanbum committed:
new created IR push to Worklist

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -278,15 +297,29 @@ bool VectorCombine::vectorizeLoadInsert(Instruction &I) {
+  SmallVector<int> Mask;
+  assert(OffsetEltIndex + VectorRange < MinVecNumElts &&
+         "Address offset too big");
+  if (!NeedCast) {
```
(style) Use if (NeedCast) {} else {} ordering

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -235,15 +237,29 @@ bool VectorCombine::vectorizeLoadInsert(Instruction &I) {
-    if (Offset.urem(ScalarSizeInBytes) != 0)
-      return false;
+    if (auto UnalignedBytes = Offset.urem(ScalarSizeInBytes);
+        UnalignedBytes != 0) {
```
Why not this?
```
if (uint64_t UnalignedBytes = Offset.urem(ScalarSizeInBytes)) {
```

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -278,15 +297,29 @@ bool VectorCombine::vectorizeLoadInsert(Instruction &I) {
+    Mask[0] = OffsetEltIndex;
+  } else {
+    Mask.assign(MinVecNumElts, PoisonMaskElem);
+    for (unsigned InsertPos = 0; InsertPos < VectorRange; InsertPos++)
```
```std::iota(Mask.begin(), Mask.begin() + VectorRange, OffsetEltIndex);```

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -295,12 +328,19 @@ bool VectorCombine::vectorizeLoadInsert(Instruction &I) {
+  Worklist.pushValue(Result);
+  if (NeedCast) {
+    Result = Builder.CreateBitOrPointerCast(Result, I.getType());
+    Worklist.pushValue(Result);
```
replaceValue should handle WorkList additions - you just need to add the instructions you created earlier

ParkHanbum committed:
remove unnecessary checks

ParkHanbum committed:
replace the for statement with std::iota and relocate the conditional statement

ParkHanbum committed:
remove duplicate pushes to Worklist

replaceValue adds new instruction to the worklist internally,
so don't need to push it to the worklist to remove it.

ParkHanbum commented:
@RKSimon I've never done that, so I'm a little worried. Can we try it first in another PR with one commit?

ParkHanbum commented:
@RKSimon  Ah! That's what you meant, I misunderstood it as something else. .

ParkHanbum commented:
@RKSimon Would you be able to review this PR when you have time?

ParkHanbum committed:
apply the revised shuffle cost calculation

ParkHanbum commented:
building in linux failed but I can't figure it out why
`[{'explained': False, 'name': 'lib/Transforms/Vectorize/CMakeFiles/LLVMVectorize.dir/VectorCombine.cpp.o', 'reason': None}]`

ParkHanbum committed:
add braces

ParkHanbum committed:
add braces

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
test coverage - add aarch64 tests?

RKSimon reviewed (commented):
Sorry this fell off my radar (keep pinging!) - hopefully we can get this finished soon

dtcxzyw reviewed (changes requested):
A correctness issue was identified by [llvm-hackme](https://github.com/dtcxzyw/llvm-hackme). Please see the [issue comment](https://github.com/llvm/llvm-project/pull/119139#issuecomment-4648109240) for details.

ParkHanbum committed:
Reject mismatched vector widths in load-insert vectorization

ParkHanbum committed:
Add mismatched bit-width regression test

ParkHanbum committed:
Add AArch64 load-insert tests

ParkHanbum committed:
Clarify the GCD chunking comment

ParkHanbum committed:
Separate scalar and chunk sizes

ParkHanbum committed:
Avoid duplicate worklist insertion

ParkHanbum committed:
Fix the out-of-range test comment

ParkHanbum committed:
Name load-insert tests by behavior

ParkHanbum committed:
Fix stale undef test comments

ParkHanbum committed:
Document vectorizeLoadInsert flow

ParkHanbum committed:
Clarify chunked load state

ParkHanbum committed:
Consolidate chunked load handling

ParkHanbum committed:
formatting
